### PR TITLE
Add a method to retrieve all points within a region to `AStarGrid2D`

### LINKED ISSUE #6276: Add a method to get all points from an AStarGrid2D
### Describe the project you are working on

turn based rpg game

### Describe the problem or limitation you are having in your project

im trying to debug and visualize what astargrid2d is doing, but i dont have a way to access all the points at once.
i realize i could just use the size + offset variables, but it would be better if we can get the underlying stored data.

PS: a get closest point function would also be very useful.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

it would be really helpful, if there was a get_all_points() function, to write debug code and visually see what the AstarGrid2D is doing

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

a new function added to the AstarGrid2D class, that returns an array of all points, something like, get_all_points() -> Array:

### If this enhancement will not be used often, can it be worked around with a few lines of script?

i think its an important addition to the AstarGrid2D class

### Is there a reason why this should be core and not an add-on in the asset library?

the AstarGrid2D class is already part of core

*Bugsquad edit (keywords for easier searching): AStar*

### PR DESCRIPTION
**Pull request by Chaosus**
Closes https://github.com/godotengine/godot-proposals/issues/6276

(Not sure whether It's useful or not)

### CODE REVIEW

**Comment by MMUTFX2053** [comment]
is there a down side to adding more utility functions ? or even classes ? you question its usefulness but i think if it helps the developers to make games and apps more easily, it should be included, unless it increases the binary size too much or slows down performance

**Comment by AThousandShips** [comment]
> is there a down side to adding more utility functions ? or even classes ?

There's such a thing as engine bloat, this feature is trivial to do yourself, the only difference is a slight loss of performance

**Comment by MMUTFX2053** [comment]
> > is there a down side to adding more utility functions ? or even classes ?
> 
> There's such a thing as engine bloat, this feature is trivial to do yourself, the only difference is a slight loss of performance

there were talks about officially maintained gdextension plugins when 4.0 hit stable, maybe an officially maintained "boost" like gdextension plugin could help offload all the bloat, and the devs still get to have tons of utility

**Comment by AThousandShips** [comment]
You couldn't really do this part as an extension, not necessarily, unless you add an extended type, you can't add methods to existing classes, but that's off topic for this PR

**Comment by Chaosus** [comment]
Sometimes its not simply to understand whether its necessary addition or not. If this considered as a bloat, I'm fine to close this PR without problem (at least its was easy to do and I need some trainings to do not forget skills).

**Comment by AThousandShips** [comment]
I think this could be useful, but the question is if it's enough of a convenience function considering the ease at which you can accomplish it in scripting 

One detail I'd suggest changing is forgoing the `_get_point_unchecked` and instead just performing the lookup manually using an offset, we already have the range so it'd be easier to replace that call and the subtractions with just iterating over an offset from the actual size

File: core/math/a_star_grid_2d.cpp [hunk 1]
```diff
@@ -558,6 +558,24 @@ Vector2 AStarGrid2D::get_point_position(const Vector2i &p_id) const {
...
+	const Rect2i inter_region = region.intersection(p_region);
+
+	const int32_t end_x = inter_region.get_end().x;
+	const int32_t end_y = inter_region.get_end().y;
+
+	Vector<Vector2> positions;
+
+	for (int32_t y = inter_region.position.y; y < end_y; y++) {
+		for (int32_t x = inter_region.position.x; x < end_x; x++) {
+			positions.push_back(_get_point_unchecked(Vector2i(x, y))->pos);
```
Comment (AThousandShips) [inline core/math/a_star_grid_2d.cpp:572]:
```suggestion
	const int32_t end_x = inter_region.get_end().x - region.position.x;
	const int32_t end_y = inter_region.get_end().y - region.position.y;

	Vector<Vector2> positions;
	positions.resize(inter_region.size.x * inter_region.size.y);
	Vector2 *ptrw = positions.ptrw();

	for (int32_t y = inter_region.position.y - region.position.y; y < end_y; y++) {
		for (int32_t x = inter_region.position.x - region.position.x; x < end_x; x++) {
			*ptrw++ = points[y][x].pos;
```
Try this perhaps, including avoiding COW stuff

**Review by AThousandShips — COMMENTED** [review]
I think the code and everything make sense, though I haven't tested it, but I'm not convinced this is necessary to put in core, I'm not sure I see the specific need to be able to fetch all the points directly, I'd probably expect any usage of this for things like drawing to just iterate over the positions and using the `get_point_position` or similar, I'm less sure about the usefulness of fetching all the points at once, especially since it doesn't consider solid points so any such checking would require further testing (which couldn't be done with this output anyway as the solid check checks for the grid points)

Also any user who might be interested in the points in any other ordering, like by column, or right to left instead etc. couldn't use this

So I'm not sure about the actual scope where this would be used directly, and the only benefit would be performance for fetching this particular case, but I'm not sure that performance would be critical if one already iterates over the output and processes it, so I'm unsure about the specific use case for this

But very interesting and nice work either way, would 100% approve if I wasn't unsure about the necessity

Edit: What I think *might* be really interesting though would be to return an array of dictionaries containing point information, like `{id, pos, solid, weight_scale}`, *that* I feel would be very useful, and would greatly help with performance etc.

File: core/math/a_star_grid_2d.cpp [hunk 2]
```diff
@@ -558,6 +558,33 @@ Vector2 AStarGrid2D::get_point_position(const Vector2i &p_id) const {
...
+	const int32_t start_x = inter_region.position.x - region.position.x;
+	const int32_t start_y = inter_region.position.y - region.position.y;
+	const int32_t end_x = inter_region.get_end().x - region.position.x;
+	const int32_t end_y = inter_region.get_end().y - region.position.y;
+
+	TypedArray<Dictionary> data;
+
+	for (int32_t y = start_y; y < end_y; y++) {
+		for (int32_t x = start_x; x < end_x; x++) {
+			const Point p = points[y][x];
```
Comment (AThousandShips) [inline core/math/a_star_grid_2d.cpp:574]:
```suggestion
			const Point &p = points[y][x];
```

**Comment by Chaosus** [comment]
@AThousandShips Yeah, it's good idea - I've changed it to return a `TypedDictionary` (with id, position, solid and weight_scale members):

изображение

**Review by AThousandShips — APPROVED** [review]
LGTM, makes sense and should be useful! 🎉

File: doc/classes/AStarGrid2D.xml [hunk 3]
```diff
@@ -106,6 +106,13 @@
 				Returns the weight scale of the point associated with the given [param id].
 			</description>
 		</method>
+		<method name="get_points_data_in_region" qualifiers="const">
```
Comment (dalexeev) [inline doc/classes/AStarGrid2D.xml:109]: Should it be called `get_point_data_in_region`?

File: doc/classes/AStarGrid2D.xml [hunk 4]
```diff
@@ -106,6 +106,13 @@
...
+			<return type="Dictionary[]" />
+			<param index="0" name="region" type="Rect2i" />
+			<description>
+				Returns an array of dictionaries with point data ([Vector2i] id, [Vector2] position, [bool] solid, [float] weight_scale) within a [param region].
```
Comment (dalexeev) [inline doc/classes/AStarGrid2D.xml:113]:
```suggestion
				Returns an array of dictionaries with point data ([code]id[/code]: [Vector2i], [code]position[/code]: [Vector2], [code]solid[/code]: [bool], [code]weight_scale[/code]: [float]) within a [param region].
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**